src/expr.cpp

```diff
@@ -9011,14 +9011,77 @@ static std::pair<std::string, const Type *> lDeduceParam(const Type *paramType,
     return std::pair<std::string, const Type *>("", nullptr);
 }
 
+bool lCheckDeductionCompatibility(TemplateInstantiation &inst, const std::pair<std::string, const Type *> &deduction,
+                                  SourcePos pos) {
+    if (deduction.second == nullptr) {
+        // Deduction failed
+        return false;
+    }
+
+    // Check if this deduction is compatible with previously deduced arguments
+    const Type *previousDeductionResult = inst.InstantiateType(deduction.first);
+
+    if (previousDeductionResult == nullptr) {
+        // This template parameter was deducted for the first time. Add it to the map.
+        inst.AddArgument(deduction.first, TemplateArg(deduction.second, pos));
+        return true;
+    }
+
+    if (!Type::Equal(previousDeductionResult, deduction.second)) {
+        if (previousDeductionResult->IsUniformType() && deduction.second->IsVaryingType() &&
+            Type::Equal(previousDeductionResult->GetAsVaryingType(), deduction.second)) {
+            // Override previous deduction with varying type
+            inst.AddArgument(deduction.first, TemplateArg(deduction.second, pos));
+            return true;
+        }
+
+        if (previousDeductionResult->IsVaryingType() && deduction.second->IsUniformType()) {
+            // That's fine, uniform will be broadcasted
+            return true;
+        }
+
+        // Deduction failed due to conflicting deduction types
+        return false;
+    }
+
+    // Deducted type is the same as previously deduced one
+    return true;
+}
+
+bool lDeduceVectorParams(TemplateInstantiation &inst, const VectorType *paramType, const VectorType *argType,
+                         SourcePos pos) {
+    const Type *elemType = paramType->GetElementType();
+    const TemplateTypeParmType *templTypeParam = CastType<TemplateTypeParmType>(elemType);
+    bool countDependent = paramType->IsCountDependent();
+    if (templTypeParam && countDependent) {
+        // Here we have param type like T<N>, and argument type like int<4>
+        // First, deduce the element type T from the argument type.
+        const Type *argElemType = argType->GetElementType();
+        auto deduction = lDeduceParam(elemType, argElemType);
+        if (!lCheckDeductionCompatibility(inst, deduction, pos)) {
+            // Deduction failed, skip to the next candidate.
+            return false;
+        }
+
+        // Then deduce the count N from the argument type.
+        // TODO!: Can argument contain not a constant count?
```

Comment:
No. As far as I remember now we support integer constants and enums only. I don't expect that we will support anything here that is is not const expression known in compile time.